File: week_1/day-2-prefix-sum-hashing/subarray_sum_k_follow_up_questions/subarray_sum_indices.js
// subarray_sum_indices.js

function subarraySumIndices(nums, k) {
    let prefixSum = 0;
    const indexMap = new Map();
    indexMap.set(0, [-1]);
    const result = [];

    for (let i = 0; i < nums.length; i++) {
        prefixSum += nums[i];

        if (indexMap.has(prefixSum - k)) {
            for (let start of indexMap.get(prefixSum - k)) {
                result.push([start + 1, i]);
            }
        }

        if (!indexMap.has(prefixSum)) {
            indexMap.set(prefixSum, []);
        }
        indexMap.get(prefixSum).push(i);
    }

    return result;
}

// Example Input
const nums = [1, 2, 3, 1, 2];
const k = 3;

// Function Call
console.log(subarraySumIndices(nums, k));
// Output: [ [0, 1], [2, 2], [3, 4] ]

File: week_1/day-2-prefix-sum-hashing/subarray_sum_k_follow_up_questions/subarray_sum_indices.py
# subarray_sum_indices.py

def subarraySumIndices(nums, k):
    prefix_sum = 0
    index_map = {0: [-1]}
    result = []

    for i, num in enumerate(nums):
        prefix_sum += num
        if prefix_sum - k in index_map:
            for start in index_map[prefix_sum - k]:
                result.append((start + 1, i))
        index_map.setdefault(prefix_sum, []).append(i)

    return result

# Example Input
nums = [1, 2, 3, 1, 2]
k = 3

# Function Call
print(subarraySumIndices(nums, k))
# Output: [(0, 1), (2, 2), (3, 4)]
